src/wifi.rs

```diff
@@ -195,6 +223,37 @@ impl<A: AtatClient, T: Timer<TIMER_HZ>, const TIMER_HZ: u32, const TX_SIZE: usiz
         let responses = self.send_command(ObtainLocalAddressCommand::new())?;
         LocalAddress::from_responses(responses)
     }
+
+    /// Restarts the module and blocks until the module is ready.
+    /// If module is not ready within five seconds, [RestartErrors::ReadyTimeout] is returned
+    fn restart(&mut self) -> Result<(), RestartErrors> {
+        self.ready = false;
+        self.send_command(RestartCommand::default())?;
+
+        self.joined = false;
+        self.ip_assigned = false;
+        self.sockets = [SocketState::Closed; 5];
+        self.data_available = [0; 5];
+        self.data = None;
+        self.passive_mode_enabled = false;
+        self.multi_connections_enabled = false;
```

Comment:
Maybe it would make sense to move the state reset into a dedicated `fn reset_state(&mut self)` method? Might be useful in the future, and would make the code slightly clearer to read.

Another observation from my experience writing stateful networking code: In most cases it's less error-prone to be able to throw away the entire state when recreating a connection, as opposed to resetting the state, and having to make sure to remember it when adding new stateful features. One implementation option might be to put all state into an inner "session" struct inside the `Adapter`, and - when re-initializing - throwing away the old struct and replacing it with a new instance. In Rust, this can be done with [mem::replace](https://doc.rust-lang.org/std/mem/fn.replace.html). This approach would ensure that every value will be reset to the defaults, and that there's no state accidentally carried over from the previous session.

Of course it's only a suggestion, feel free to ignore 🙂